File: HW7/PolygonVect.java
import java.util.ArrayList;
import java.awt.Graphics;
import java.awt.*;
/**
* @author Goktug Akin
* @since 2019
* @version HW7
*/
public class PolygonVect extends Polygon{

	private ArrayList<Point2D> _points; 

	/**
	* Constructor that accpets 2D points by ArrayList.
	* @param points 2D points with contained by ArrayList.
	* @throws NegativeArraySizeException If parameter points has illegal size (less or equal than zero).
	*/
	public PolygonVect(ArrayList<Point2D> points)throws NegativeArraySizeException
	{
		if(points.size()<=0)
			throw new NegativeArraySizeException("Invalid size passing PolygonVect(ArrayList<Point2D> ..) constructor");
		_points = new ArrayList<Point2D>();
		for (int i=0;i<points.size();++i)
			_points.add(points.get(i));
		
	}
	/**
	* Conversion constructor converts from <b>Rectangle</b>
	* @param rect Rectangle object that will be converted to Polygon.
	*/
	public PolygonVect(Rectangle rect)
	{	
		_points = new ArrayList<Point2D>();
		Point2D temp=new Point2D(rect.get_cx(),rect.get_cy());
		_points.add(temp);
		Point2D temp2=new Point2D(rect.get_cx()+rect.get_width(),rect.get_cy());
		_points.add(temp2);
		Point2D temp3=new Point2D(rect.get_cx()+rect.get_width(),rect.get_cy()+rect.get_height());
		_points.add(temp3);
		Point2D temp4=new Point2D(rect.get_cx(),rect.get_cy()+rect.get_height());
		_points.add(temp4);
	}
	/**
	* Conversion constructor converts from <b>Triangle</b>
	* @param tri Triangle object that will be converted to Polygon.
	*/
	public PolygonVect(Triangle tri)
	{	
		_points = new ArrayList<Point2D>();
		Point2D temp=new Point2D(tri.get_cr1()[0],tri.get_cr1()[1]);
		_points.add(temp);
		Point2D temp2=new Point2D(tri.get_cr2()[0],tri.get_cr2()[1]);
		_points.add(temp2);
		Point2D temp3=new Point2D(tri.get_cr3()[0],tri.get_cr3()[1]);
		_points.add(temp3);
	}
	/**
	* Conversion constructor converts from <b>Circle</b>
	* @param circ Circle object that will be converted to Polygon.
	*/
	public PolygonVect(Circle circ)
	{	
		_points = new ArrayList<Point2D>();
		Point2D temp = new Point2D();
		double angle = 0;
		for (int i = 0; i < 100; ++i)
		{
			temp=new Point2D(circ.get_radius()*Math.cos(angle)+circ.get_cx(),circ.get_radius()*Math.sin(angle)+circ.get_cy());
			angle+=3.6*(3.14159/180);//For 100 equal intervals angle changes
			_points.add(temp);
		}
	}
	/**
	* Conversion constructor converts from <b>PolygonDyn</b>
	* @param pdyn PolygonDyn object that will be converted to Polygon.
	* @throws NegativeArraySizeException If param pdyn has illegal number of points (less or equal than zero).
	*/
	public PolygonVect(PolygonDyn pdyn)throws NegativeArraySizeException
	{
		if(pdyn.size()<=0)
			throw new NegativeArraySizeException("Invalid size passed to PolygonVect(PolygonDyn pdyn) constructor.");
		else
		{	
			_points = new ArrayList<Point2D>();
			for (int i=0;i<this.size();++i)
				_points.remove(i);
			for (int i=0;i<pdyn.size();++i)
				_points.add(pdyn.getPoint(i));
		}
	}
	/**
	* Retrieves the number of the points.
	* @return Number of Point2D objects. 
	*/
	public int size()
	{
		return _points.size();
	}
	/**
	* Retrieves the point with in the given index.
	* @param index index value for 2D point
	* @return Point2D object with given index.
	* @throws NegativeArraySizeException If index is illegal (less than zero).
	*/
	public Point2D getPoint(int index)throws NegativeArraySizeException
	{	
		if(index<0)
			throw new NegativeArraySizeException("Invalid index passed to getPoint(int index) for PolygonVect.");
		else
			return _points.get(index);
	}
	/**
	* Retrieves the area of the polygon.
	* @return Area of the polygon.
	*/
	@Override
	public double area()
	{
		// Calculating area for coordinate geometry
		// Source : https://www.mathopenref.com/coordpolygonarea.html
		double temp = 0;
		for (int i = 0,j=0; i < _points.size() ; ++i,++j){
			if(i==_points.size()-1)
				j=-1;
			temp+=((_points.get(i).get_x()*_points.get(j+1).get_y()) - (_points.get(i).get_y()*_points.get(j+1).get_x()));	
		}
		if(temp<0)
			temp=temp*-1;
		
		return temp / 2;
	}
	/**
	* Retrieves the perimeter of the polygon.
	* @return Perimeter of the polygon.
	*/
	@Override
	public double perimeter()
	{	
		// Calculating perimeter for coordinate geometry
		// Source : https://www.mathopenref.com/coordpolygonarea.html
		double temp = 0;
		for (int i = 0,j=0; i < _points.size();++i,++j)
		{
			if(i==_points.size()-1)
				j=-1;
			temp+= Math.sqrt((Math.pow(_points.get(i).get_x() - _points.get(j+1).get_x(),2)) + (Math.pow(_points.get(i).get_y() - _points.get(j+1).get_y(),2)));
			
		}
	 	return temp;
	}

	/**
	* Draw polygon with graphics object.
	* @param g Graphics object.
	*/
	@Override
	public void draw(Graphics g)
	{

		int [] x_points = new int[_points.size()];
		int [] y_points = new int[_points.size()];

		for (int i=0;i<_points.size();++i) 
			x_points[i]=(int)_points.get(i).get_x();
		for (int i=0;i<_points.size();++i) 
			y_points[i]=(int)_points.get(i).get_y();
		
		g.fillPolygon(x_points,y_points,_points.size());
		g.setColor(Color.black);
        for (int i=0; i<_points.size(); ++i)
            x_points[i]--;
        for(int j=0; j<_points.size(); ++j)
            y_points[j]--;
        g.drawPolygon(x_points,y_points,_points.size()); 
	}
	/**
	* Increments all of the coordinates of polygon.
	*/
	@Override
	public void increment()
	{
		double x,y;
		for (int i=0;i<_points.size();++i) {
			x=_points.get(i).get_x();
			x--;
			y=_points.get(i).get_y();
			y--;
			_points.get(i).set_points(x,y);		
		}
	}
	/**
	* Decrements all of the coordinates of polygon.
	*/
	@Override
	public void decrement()
	{
		double x,y;
		for (int i=0;i<_points.size();++i) {
			x=_points.get(i).get_x();
			x++;
			y=_points.get(i).get_y();
			y++;
			_points.get(i).set_points(x,y);		
		}
	}

	@Override
	public String toString()
	{
		return String.format("PolygonVect with %d points", _points.size());
	}

}

File: HW5/polygonvect.cpp
#include "polygonvect.h"
#include <cmath>

namespace gtuShapes
{	
	polygonvect::polygonvect(std::vector<Point2D> points)throw(invalidsize) : polygon() 
	{	
		if(points.size()==0)
			throw invalidsize();
		for (unsigned int i = 0; i < points.size(); ++i)
			_points.push_back(points[i]);
		this->setColor("green");	
	}
	polygonvect::polygonvect(shape * shptr)throw(invalidptr) : polygon() 
	{
		/* This constructor takes shape pointer, determine its type and converts 
		I know I am duplicating the code but, I cant use delegating constructor
		(to prevent duplicate) inside body of the constructor */

		rectangle * isRectangle = dynamic_cast<rectangle*>(shptr);
		triangle * isTriangle = dynamic_cast<triangle*>(shptr);
		circle * isCircle = dynamic_cast<circle*>(shptr);

		if(isRectangle!=nullptr)
		{
			polygon::Point2D temp;
			temp.set_points(isRectangle->get_cx(),isRectangle->get_cy());
			_points.push_back(temp);
			temp.set_points(isRectangle->get_cx()+isRectangle->get_width(),isRectangle->get_cy());
			_points.push_back(temp);
			temp.set_points(isRectangle->get_cx()+isRectangle->get_width(),isRectangle->get_cy()+isRectangle->get_height());
			_points.push_back(temp);
			temp.set_points(isRectangle->get_cx(),isRectangle->get_cy()+isRectangle->get_height());
			_points.push_back(temp);
		}
		else if(isTriangle!=nullptr)
		{
			polygon::Point2D temp;
			temp.set_points(isTriangle->get_cr1()[0],isTriangle->get_cr1()[1]);
			_points.push_back(temp);
			temp.set_points(isTriangle->get_cr2()[0],isTriangle->get_cr2()[1]);
			_points.push_back(temp);
			temp.set_points(isTriangle->get_cr3()[0],isTriangle->get_cr3()[1]);
			_points.push_back(temp);
		}
		else if(isCircle!=nullptr)
		{
			polygon::Point2D temp;
			double angle = 0;
			for (int i = 0; i < 100; ++i)
			{
				temp.set_points(isCircle->get_radius()*cos(angle)+isCircle->get_cx(),isCircle->get_radius()*sin(angle)+isCircle->get_cy());
				angle+=3.6*(3.14159/180);//For 100 equal intervals angle changes
				_points.push_back(temp);
			}
		}
		else
			throw invalidptr();
	}
	polygonvect::polygonvect(const triangle& tri) : polygon()
	{	
		polygon::Point2D temp;
		temp.set_points(tri.get_cr1()[0],tri.get_cr1()[1]);
		_points.push_back(temp);
		temp.set_points(tri.get_cr2()[0],tri.get_cr2()[1]);
		_points.push_back(temp);
		temp.set_points(tri.get_cr3()[0],tri.get_cr3()[1]);
		_points.push_back(temp);
	}
	polygonvect::polygonvect(const rectangle& rect) : polygon()
	{	
		polygon::Point2D temp;
		temp.set_points(rect.get_cx(),rect.get_cy());
		_points.push_back(temp);
		temp.set_points(rect.get_cx()+rect.get_width(),rect.get_cy());
		_points.push_back(temp);
		temp.set_points(rect.get_cx()+rect.get_width(),rect.get_cy()+rect.get_height());
		_points.push_back(temp);
		temp.set_points(rect.get_cx(),rect.get_cy()+rect.get_height());
		_points.push_back(temp);

	}
	polygonvect::polygonvect(const circle& circ) : polygon()
	{	
		polygon::Point2D temp;
		double angle = 0;
		for (int i = 0; i < 100; ++i)
		{
			temp.set_points(circ.get_radius()*cos(angle)+circ.get_cx(),circ.get_radius()*sin(angle)+circ.get_cy());
			angle+=3.6*(3.14159/180);//For 100 equal intervals angle changes
			_points.push_back(temp);
		}

	}
	const polygon::Point2D& polygonvect::operator[](int index)const throw(outofrange)
	{	
		if (index<0)
		{
			throw outofrange();
		}
		return _points[index];
	}
	polygon::Point2D& polygonvect::operator[](int index) throw (outofrange)
	{
		if (index<0)
		{
			throw outofrange();
		}
		return _points[index];
	}
	
	double polygonvect::area()const
	{	
		// Calculating area for coordinate geometry
		// Source : https://www.mathopenref.com/coordpolygonarea.html
		double temp = 0;
		for (unsigned int i = 0,j=0; i < _points.size() ; ++i,++j){
			if(i==_points.size()-1)
				j=-1;
			temp+=((_points[i].get_x()*_points[j+1].get_y()) - (_points[i].get_y()*_points[j+1].get_x()));	
		}
		if(temp<0)
			temp=temp*-1;
		
		return temp / 2;
	}
	double polygonvect::perimeter()const
	{	
		// Calculating perimeter for coordinate geometry
		// Source : https://www.mathopenref.com/coordpolygonarea.html
		double temp = 0;
		for (unsigned int i = 0,j=0; i < _points.size();++i,++j)
		{
			if(i==_points.size()-1)
				j=-1;
			temp+= sqrt((pow(_points[i].get_x() - _points[j+1].get_x(),2)) + (pow(_points[i].get_y() - _points[j+1].get_y(),2)));
			
		}
	 	return temp;
	 }
	const shape& polygonvect::operator++()
	{
		double tmpx;
		double tmpy;
		for (unsigned int i = 0; i < _points.size(); ++i)
		{
			tmpx = _points[i].get_x() + 1;
			tmpy = _points[i].get_y() + 1 ;
			_points[i].set_points(tmpx,tmpy);
		}
		return *this;
	}
	
    const shape& polygonvect::operator++(int ignore) throw(notavailable)
    {
    	throw notavailable();
    }
	const shape& polygonvect::operator--()
	{
		double tmpx;
		double tmpy;
		for (unsigned int i = 0; i < _points.size(); ++i)
		{
			tmpx = _points[i].get_x() - 1;
			tmpy = _points[i].get_y() - 1 ;
			_points[i].set_points(tmpx,tmpy);
		}
		return *this;
	}
	 const shape& polygonvect::operator--(int ignore) throw(notavailable)
    {
    	throw notavailable();
    }
    std::ostream& polygonvect::print(std::ostream& outputStream)const
    {
    	return outputStream << *this;
    }
    std::ostream& operator<<(std::ostream& outputStream, const polygonvect& p1)
    {
		outputStream << "<polygon points=\"";
		for (unsigned int i = 0; i < p1._points.size(); ++i)
			outputStream << p1._points[i].get_x() << "," << p1._points[i].get_y() << " " ;

		outputStream << "\" stroke=\"black\" stroke-width=\"2\" fill=\"" 
					<< p1.getColor() << "\" />" << std::endl;
		return outputStream;
    }
}